packages/mobx-state-tree/src/core/node/object-node.ts

```diff
@@ -20,73 +19,107 @@ import {
     createActionInvoker,
     NodeLifeCycle,
     IdentifierCache,
-    EMPTY_ARRAY,
     escapeJsonPath,
     addHiddenFinalProp,
     toJSON,
     freeze,
-    resolveNodeByPathParts
+    resolveNodeByPathParts,
+    convertChildNodesToArray,
+    ModelType
 } from "../../internal"
 
 let nextNodeId = 1
 
+export interface IChildNodesMap {
+    [key: string]: INode
+}
+
 export class ObjectNode implements INode {
     nodeId = ++nextNodeId
     readonly type: IType<any, any>
-    readonly storedValue: any
+    storedValue: any
     @observable subpath: string = ""
     @observable protected _parent: ObjectNode | null = null
     _isRunningAction = false // only relevant for root
 
     identifierCache: IdentifierCache | undefined
     isProtectionEnabled = true
-    identifierAttribute: string | undefined = undefined // not to be modified directly, only through model initialization
+    readonly identifierAttribute: string | undefined
     _environment: any = undefined
     protected _autoUnbox = true // unboxing is disabled when reading child nodes
     state = NodeLifeCycle.INITIALIZING
 
-    middlewares = EMPTY_ARRAY as IMiddleware[]
-    private snapshotSubscribers: ((snapshot: any) => void)[]
-    private patchSubscribers: ((patch: IJsonPatch, reversePatch: IJsonPatch) => void)[]
-    private disposers: (() => void)[]
+    middlewares: IMiddleware[] | null = null
+    private snapshotSubscribers: ((snapshot: any) => void)[] | null = null
+    private patchSubscribers:
+        | ((patch: IJsonPatch, reversePatch: IJsonPatch) => void)[]
+        | null = null
+    private disposers: (() => void)[] | null = null
 
-    applyPatches: (patches: IJsonPatch[]) => void
+    applyPatches(patches: IJsonPatch[]): void {
+        if (!this._observableInstanceCreated) this._createObservableInstance()
+        this.applyPatches(patches)
+    }
+    applySnapshot(snapshot: any): void {
+        if (!this._observableInstanceCreated) this._createObservableInstance()
+        this.applySnapshot(snapshot)
+    }
 
-    applySnapshot: (snapshot: any) => void
+    private _observableInstanceCreated: boolean = false
+    private readonly _childNodes: IChildNodesMap | null = null
```

Comment:
I was wondering about this one, if the instance is not yet created, that means that all children are still (in serialized form probably) in that initial snapshot value, so none of the children need or should be constructed either?

Reply:
If I understood the question correctly, this was the original idea to "end up creating just the **entire Node tree** instead of the whole observable tree" ([comment 1](https://github.com/mobxjs/mobx-state-tree/issues/810#issuecomment-388020137), [comment 2 ](https://github.com/mobxjs/mobx-state-tree/issues/810#issuecomment-390393224));

Initializing children as ObjectNodes (without observable counterpart) is needed to:
- fill identifierCache - otherwise references won't work;
- getting deeply nested children for ```snapshot```: root is fully initialized => ```.snapshot``` calls ```getActualSnapshot()``` => it goes to ```type.getSnapshot()``` => for any complex type it goes through ```this.getChildNode(node, name)``` => if child is not yet fully initiazed, it goes to ```this._childNodes![subpath]```
- finalizeCreation() uses getChildren() to fire hooks, which is expected to return nodes, not just objects

I could forget a point or two, becuase I had no chance to peek into the code.

Reply:
You are correct, thanks!